vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,876 @@
+void automl<CMType>::offset_learn(multi_learner& base, multi_ex& ec, CB::cb_class& logged, size_t labelled_action)
+{
+  for (size_t live_slot = 0; live_slot < cm->scores.size(); ++live_slot)
+  {
+    for (example* ex : ec) { cm->apply_config(ex, live_slot); }
+
+    auto restore_guard = VW::scope_exit([this, &ec, &live_slot] {
+      for (example* ex : ec) { this->cm->revert_config(ex); }
+    });
+
+    if (!base.learn_returns_prediction) { base.predict(ec, live_slot); }
+
+    base.learn(ec, live_slot);
+
+    const auto& action_scores = ec[0]->pred.a_s;
+    // cb_adf => first action is a greedy action
+    const auto maxit = action_scores.begin();
+    const uint32_t chosen_action = maxit->action;
+
+    // extra asserts
+    assert(chosen_action < ec.size());
+    assert(labelled_action < ec.size());
+
+    const float w = logged.probability > 0 ? 1 / logged.probability : 0;
+    const float r = -logged.cost;
+
+    cm->scores[live_slot].update(chosen_action == labelled_action ? w : 0, r);
+
+    // cache the champ
+    if (cm->current_champ == live_slot) { champ_a_s = std::move(ec[0]->pred.a_s); }
+  }
+  // replace bc champ always gets cached
+  ec[0]->pred.a_s = std::move(champ_a_s);
+}
+
+// this is the registered learn function for this reduction
+// mostly uses config_manager and actual_learn(..)
+template <typename CMType, bool is_explore>
+void learn_automl(automl<CMType>& data, multi_learner& base, multi_ex& ec)
+{
+  assert(data.all->weights.sparse == false);
+
+  bool is_learn = true;
+
+  // extra assert just bc
+  assert(data.all->interactions.empty() == true);
+
+  CB::cb_class logged{};
+  size_t labelled_action = 0;
+  if (is_learn)
+  {
+    const auto it = std::find_if(ec.begin(), ec.end(), [](example* item) { return !item->l.cb.costs.empty(); });
+
+    if (it != ec.end())
+    {
+      logged = (*it)->l.cb.costs[0];
+      labelled_action = std::distance(ec.begin(), it);
+    }
+  }
+
+  data.one_step(base, ec, logged, labelled_action);
+
+  // extra: assert again just like at the top
+  assert(data.all->interactions.empty() == true);
+  assert(ec[0]->interactions == nullptr);
+}
+
+template <typename CMType>
+void persist(automl<CMType>& data, metric_sink& metrics)
+{
+  data.cm->persist(metrics);
+}
+
+template <typename CMType>
+void finish_example(vw& all, automl<CMType>& data, multi_ex& ec)
+{
+  {
+    size_t champ_live_slot = data.cm->current_champ;
+    for (example* ex : ec) { data.cm->apply_config(ex, champ_live_slot); }
+
+    auto restore_guard = VW::scope_exit([&data, &ec, &champ_live_slot] {
+      for (example* ex : ec) { data.cm->revert_config(ex); }
+    });
+
+    data.adf_learner->print_example(all, ec);
+  }
+
+  VW::finish_example(all, ec);
+}
+
+template <typename CMType>
+void save_load_aml(automl<CMType>& d, io_buf& model_file, bool read, bool text)
+{
+  if (model_file.num_files() == 0) { return; }
+  VW::model_utils::process_model_field(model_file, d.current_state, read, "_aml_state", text);
```

Comment:
Some encapsulation here might make sense:
auto_ml.save_load()
  if(model...
  process_model_file(...)
  cm->save_load(..)

Reply:
Im planning on changing all the save_load stuff once this is merged. Check out this PR to see the direction:
https://github.com/VowpalWabbit/vowpal_wabbit/pull/3358
It should all get a lot cleaner